app/vmagent/datadog/request_handler.go

```diff
@@ -28,22 +28,35 @@ func InsertHandlerForHTTP(at *auth.Token, req *http.Request) error {
 		return err
 	}
 	return stream.Parse(
-		req, func(series prompbmarshal.TimeSeries) error {
-			series.Labels = append(series.Labels, extraLabels...)
-			return insertRows(at, series)
+		req, func(series []prompbmarshal.TimeSeries) error {
+			return insertRows(at, series, extraLabels)
 		},
 	)
 }
 
-func insertRows(at *auth.Token, series prompbmarshal.TimeSeries) error {
+func insertRows(at *auth.Token, series []prompbmarshal.TimeSeries, extraLabels []prompbmarshal.Label) error {
 	ctx := common.GetPushCtx()
 	defer common.PutPushCtx(ctx)
 
-	rowsTotal := len(series.Samples)
+	var rowsTotal int
+	tssDst := ctx.WriteRequest.Timeseries[:0]
+	labels := ctx.Labels[:0]
+	samples := ctx.Samples[:0]
 
-	ctx.WriteRequest.Timeseries = []prompbmarshal.TimeSeries{series}
-	ctx.Labels = series.Labels
-	ctx.Samples = series.Samples
+	for i := range series {
+		s := &series[i]
+		rowsTotal += len(s.Samples)
+
+		s.Labels = append(s.Labels, extraLabels...)
```

Comment:
I think, intention of using labels := ctx.Labels[:0] - reduce memory usage with shared memory, Timseries.Labeles is actually a reference to ctx.Labels. And take an ownership of timeseries.


```go
// save pointer to the current memory end:
dstLabelsLen := len(labels)
samplesLen := len(samples)
// copy labels into shared memory
labels = append(labels, s.Labels...)
// copy extra labels
labels = append(labels, extraLabels...)
// copy samples to shared memory
samples = append(samples, s.Samples...)
// copy memory, we shouldn't hold a reference for timeseries
tssDst = append(tssDst, prompbmarshal.TimeSeries{
    Labels: labels[dstLabelsLen:],
    Samples: samples[samplesLen:],
 })
```

Reply:
I'm not sure that ctx.Labels is used as a shared memory across multiple series.
According to [the code](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/e1359c904c75db18716918eee93e42d6357b2b1e/app/vmagent/common/push_ctx.go#L15-L16) ctx.Labels still need to contain flat list of all the labels in request.
So in this code I do the following:
1. update prompbmarshal.TimeSeries with extra labels
2. add all labels from current prompbmarshal.TimeSeries to flatlist ctx.Labels

prompbmarshal.TimeSeries comes as a copy to this func, so looks like I can reference it without consequences.

Reply:
In fact, it is. It was a part of previous implementation. 

 For instance:

```go
// allocate memory and copy - s.Labels do not re-used and slice may be extended via Grow call.
s.Labels = append(s.Labels, extraLabels...)
// do not allocate memory and copy. Labels re-used and there is no need at Grow call
labels =  append(labels, extraLabels...)
```

Reply:
Any way, @valyala ll re-write it...